drivers/ethernet/eth_xlnx_gem.c

```diff
@@ -0,0 +1,1724 @@
+		/* [24] Discard RX packet when AHB unavailable */
+		reg_val |= ETH_XLNX_GEM_DMACR_DISCNOAHB_BIT;
+	}
+	/*
+	 * [23..16] DMA RX buffer size in AHB system memory
+	 *    e.g.: 0x02 = 128, 0x18 = 1536, 0xA0 = 10240
+	 */
+	reg_val |= (
+		((dev_conf->rx_buffer_size / 64)
+		& ETH_XLNX_GEM_DMACR_RX_BUF_MASK)
+		<< ETH_XLNX_GEM_DMACR_RX_BUF_SHIFT);
+	if (dev_conf->enable_tx_chksum_offload == 1) {
+		/* [11] TX TCP/UDP/IP checksum offload to GEM */
+		reg_val |= ETH_XLNX_GEM_DMACR_TCP_CHKSUM_BIT;
+	}
+	if (dev_conf->tx_buffer_size_full == 1) {
+		/* [10] TX buffer memory size select */
+		reg_val |= ETH_XLNX_GEM_DMACR_TX_SIZE_BIT;
+	}
+	/*
+	 * [09..08] RX packet buffer memory size select
+	 *          0 = 1kB, 1 = 2kB, 2 = 4kB, 3 = 8kB
+	 */
+	reg_val |= (
+		((uint32_t)dev_conf->hw_rx_buffer_size
+		<< ETH_XLNX_GEM_DMACR_RX_SIZE_SHIFT)
+		& ETH_XLNX_GEM_DMACR_RX_SIZE_MASK);
+	if (dev_conf->enable_ahb_packet_endian_swap == 1) {
+		/* [07] AHB packet data endian swap enable */
+		reg_val |= ETH_XLNX_GEM_DMACR_ENDIAN_BIT;
+	}
+	if (dev_conf->enable_ahb_md_endian_swap == 1) {
+		/* [06] AHB mgmt descriptor endian swap enable */
+		reg_val |= ETH_XLNX_GEM_DMACR_DESCR_ENDIAN_BIT;
+	}
+	/*
+	 * [04..00] AHB fixed burst length for DMA ops.
+	 *          00001 = single AHB bursts,
+	 *          001xx = attempt to use INCR4  bursts,
+	 *          01xxx = attempt to use INCR8  bursts,
+	 *          1xxxx = attempt to use INCR16 bursts
+	 */
+	reg_val |= ((uint32_t)dev_conf->ahb_burst_length
+		& ETH_XLNX_GEM_DMACR_AHB_BURST_LENGTH_MASK);
+
+	/* Write the assembled register contents */
+	sys_write32(reg_val, dev_conf->base_addr + ETH_XLNX_GEM_DMACR_OFFSET);
+}
+
+/**
+ * @brief GEM associated PHY detection and setup function
+ * If the current GEM device shall manage an associated PHY, its detection
+ * and configuration is performed from within this function. Called from
+ * within the device initialization function. This function refers to
+ * functionality implemented in the phy_xlnx_gem module.
+ *
+ * @param dev Pointer to the device data
+ */
+static void eth_xlnx_gem_init_phy(const struct device *dev)
+{
+	struct eth_xlnx_gem_dev_data *dev_data = DEV_DATA(dev);
+	int detect_rc = 0;
+
+	LOG_DBG("%s attempting to initialize associated PHY",
+		dev->name);
+
+	/*
+	 * The phy_xlnx_gem_detect function checks if a valid PHY
+	 * ID is returned when reading the corresponding high / low
+	 * ID registers for all valid MDIO addresses. If a compatible
+	 * PHY is detected, the function writes a pointer to the
+	 * vendor-specific implementations of the PHY management
+	 * functions to the run-time device data struct, along with
+	 * the ID and the MDIO address of the detected PHY (dev_data->
+	 * phy_id, dev_data->phy_addr, dev_data->phy_access_api).
+	 */
+	detect_rc = phy_xlnx_gem_detect(dev);
+
+	if (detect_rc == 0 &&
+		dev_data->phy_id != 0x00000000 &&
+		dev_data->phy_id != 0xFFFFFFFF &&
+		dev_data->phy_access_api != NULL) {
```

Comment:
Minor nit, if you indent these one or more steps right, then it is easier to read (because the block below is not in the same level).

Reply:
I've adjusted the indentation for this code block.